src/main.rs

```diff
@@ -48,24 +48,32 @@ fn main() {
     let args = Args::parse();
     let (nifti_header, data) = read_3d_image::<_, f32>(args.input_image);
 
-    let streamlines = args.fibers.map(|path| {
-        let spacing = Vector3::new(
-            nifti_header.pixdim[1],
-            nifti_header.pixdim[2],
-            nifti_header.pixdim[3],
-        );
-        let mut reader = Reader::new(path).unwrap().to_voxel_space(spacing);
-        reader.streamlines()
-    });
-
+    let streamlines: Vec<Polyline> = match args.fibers {
+        Some(path) => {
+            let spacing = Vector3::new(
+                nifti_header.pixdim[1],
+                nifti_header.pixdim[2],
+                nifti_header.pixdim[3],
+            );
+            let mut reader = Reader::new(path).unwrap().to_voxel_space(spacing);
+            reader
+                .streamlines()
+                .iter()
+                .map(|streamline| streamline.iter().map(|p| vec3(p.x, p.y, p.z)).collect())
+                .collect()
+        }
+        None => vec![],
+    };
     let slicer = Slicer::from_3d(nifti_header, data, 3, &args.views, (0.3, 0.7));
+
+    let inputs = ContextInputs {
+        streamlines,
+        dst_img_size: uvec2(args.output_size[0], args.output_size[1]),
+    };
+    let mut graphics = graphics::Context::new(&inputs);
+
     for slice in slicer.slices {
-        let inputs = UserInputs {
-            src_img: slice.data,
-            streamlines: &streamlines,
-            dst_img_size: uvec2(args.output_size[0], args.output_size[1]),
-        };
-        let image = graphics::run_full_pipeline(&inputs);
+        let image = graphics.process_slice(&slice.data);
```

Comment:
That's what I did in vtk back in the days when I realized that building the whole scene over and over was really slow. Now I'm only changing the slice. I guess that makes sense in `wgpu` too!